File: Saver/PolaroidRenderer.swift
import ScreenSaver
import AppKit

private func hand(_ size: CGFloat) -> NSFont {
    NSFont(name: "Bradley Hand", size: size)
        ?? NSFont(name: "Marker Felt", size: size)
        ?? .systemFont(ofSize: size, weight: .medium)
}

extension AlbumCoverSaverView {

    /// The board is pinned once, when the screen saver starts. After that the
    /// only thing that changes is that the record which just finished gets pinned
    /// up alongside the rest, and the new one takes the middle.
    func buildPolaroidBoard() {
        claimTable(for: .polaroid)
        guard !albums.isEmpty else { return }
        let cap = polaroidCap
        guard cap > 0 else { table.removeAll(); return }
        guard table.isEmpty else { return }

        // Skip whatever is playing — that one belongs in the centre, not the wall.
        let centre = liveAlbumIndex ?? featIndex
        for album in albums.prefix(cap + 1)
            where !(albums.indices.contains(centre) && albums[centre].id == album.id) {
            if table.count >= cap { break }
            pinToTable(album.id, spot: polaroidSpot())
        }
    }

    private var polaroidCap: Int {
        isPreview ? 4 : max(0, min(24, settings.polaroidCount))
    }

    /// Keeps clear of the middle, where the current photo hangs.
    private func polaroidSpot() -> (x: CGFloat, y: CGFloat, size: CGFloat)? {
        let W = bounds.width, H = bounds.height
        return findTableSpot(sizeRange: 0.17...0.26, scale: 1,
                             keepOutCentre: NSPoint(x: W * 0.5, y: H * 0.5),
                             keepOutRadius: min(H * 0.36, W * 0.26) * 0.82,
                             keepOutRect: nil)
    }

    func advancePolaroid() {
        advanceFeatured(idleSpan: settings.featureSeconds)
        if table.isEmpty { buildPolaroidBoard() }

        // Track the centre by album id straight off the playback signal, not off
        // the featured index. `advanceFeatured` freezes that index for 1.1s while
        // the swap animates, so a quick skip during that window used to vanish
        // without ever being pinned.
        let currentID: String?
        if nowPlaying.isLive, !nowPlaying.albumID.isEmpty {
            currentID = nowPlaying.albumID
        } else if albums.indices.contains(featIndex) {
            currentID = albums[featIndex].id
        } else {
            currentID = nil
        }
        guard let currentID else { return }

        if let previous = lastCentreAlbumID, previous != currentID {
            pinToTable(previous, spot: polaroidSpot())
            while table.count > polaroidCap { table.removeFirst() }
        }
        lastCentreAlbumID = currentID
    }

    /// Instant photos pinned to cork, swaying very slightly on their pins.
    func drawPolaroid() {
        let W = bounds.width, H = bounds.height
        let pal = featPalette

        // Wooden frame, then the cork inside it.
        let frameW = min(W, H) * 0.028
        NSGradient(starting: NSColor(red: 0.42, green: 0.27, blue: 0.14, alpha: 1),
                   ending: NSColor(red: 0.26, green: 0.16, blue: 0.08, alpha: 1))?
            .draw(in: NSBezierPath(rect: bounds), angle: -70)
        for i in 0..<40 {                       // frame grain
            let y = CGFloat(i) / 40 * H
            NSColor(white: 0, alpha: 0.05 + CGFloat(i % 3) * 0.02).setStroke()
            let l = NSBezierPath()
            l.move(to: NSPoint(x: 0, y: y)); l.line(to: NSPoint(x: W, y: y))
            l.lineWidth = 1.5
            l.stroke()
        }

        let board = bounds.insetBy(dx: frameW, dy: frameW)
        NSGraphicsContext.saveGraphicsState()
        NSBezierPath(rect: board).addClip()
        CorkTexture.image(size: board.size)
            .draw(in: board, from: .zero, operation: .sourceOver, fraction: 1)
        // The frame casts a shadow onto the cork, which is what gives the board
        // depth rather than looking like a printed border.
        for edge in 0..<4 {
            let d = frameW * 0.9
            let r: NSRect
            let angle: CGFloat
            switch edge {
            case 0: r = NSRect(x: board.minX, y: board.maxY - d, width: board.width, height: d); angle = -90
            case 1: r = NSRect(x: board.minX, y: board.minY, width: board.width, height: d); angle = 90
            case 2: r = NSRect(x: board.minX, y: board.minY, width: d, height: board.height); angle = 0
            default: r = NSRect(x: board.maxX - d, y: board.minY, width: d, height: board.height); angle = 180
            }
            NSGradient(colors: [NSColor(white: 0, alpha: 0.30), NSColor(white: 0, alpha: 0)])?
                .draw(in: NSBezierPath(rect: r), angle: angle)
        }
        NSGraphicsContext.restoreGraphicsState()

        let lip = NSBezierPath(rect: board)
        lip.lineWidth = 2
        NSColor(white: 0, alpha: 0.35).setStroke()
        lip.stroke()

        let W2 = bounds.width, H2 = bounds.height
        for (i, t) in table.enumerated() {
            guard let idx = albumIndexByID[t.albumID] else { continue }
            // Newly pinned photos settle in rather than appearing instantly.
            let age = phase - t.addedAt
            let k = age < 0.6 ? Ease.back(CGFloat(age / 0.6)) : 1
            let s = Sleeve(centre: NSPoint(x: t.nx * W2, y: t.ny * H2 + (1 - k) * H2 * 0.025),
                           side: t.sizeFrac * H2 * (1.2 - 0.2 * k),
                           angle: t.angle, index: idx, discPeek: 0, discDirection: 0)
            let ctx = NSGraphicsContext.current!.cgContext
            ctx.saveGState()
            ctx.setAlpha(k)
            drawPhoto(s, i: i, big: false, pal: pal)
            ctx.restoreGState()
        }

        guard albums.indices.contains(featIndex) else { return }
        // The current album gets the big photo, front and centre.
        let side = min(H * 0.36, W * 0.26)
        drawPhoto(Sleeve(centre: NSPoint(x: W * 0.5, y: H * 0.5), side: side,
                         angle: -0.03, index: featIndex, discPeek: 0, discDirection: 0),
                  i: 999, big: true, pal: pal)
    }

    private func drawPhoto(_ s: Sleeve, i: Int, big: Bool, pal: ArtPalette) {
        guard albums.indices.contains(s.index) else { return }
        let ctx = NSGraphicsContext.current!.cgContext
        // Photos hang from a single pin, so they rock rather than slide.
        let sway = CGFloat(sin(Double(phase) * 0.6 + Double(i) * 1.7)) * 0.012
        let w = s.side
        let border = w * 0.055
        let h = w + border * 4.4          // instant film's deep bottom margin

        ctx.saveGState()
        ctx.translateBy(x: s.centre.x, y: s.centre.y + h * 0.42)
        ctx.rotate(by: s.angle + sway)
        ctx.translateBy(x: 0, y: -h * 0.42)

        let card = NSRect(x: -w / 2 - border, y: -h / 2, width: w + border * 2, height: h)
        NSGraphicsContext.saveGraphicsState()
        let sh = NSShadow()
        sh.shadowColor = NSColor(white: 0, alpha: 0.55)
        sh.shadowBlurRadius = w * 0.11
        sh.shadowOffset = NSSize(width: 0, height: -w * 0.035)
        sh.set()
        NSColor(red: 0.97, green: 0.96, blue: 0.93, alpha: 1).setFill()
        NSBezierPath(rect: card).fill()
        NSGraphicsContext.restoreGraphicsState()

        let photo = NSRect(x: -w / 2, y: card.maxY - border - w, width: w, height: w)
        drawCover(s.index, in: photo)
        if !big {
            pal.deep.withAlphaComponent(0.20).setFill()
            NSBezierPath(rect: photo).fill()
        }

        let album = albums[s.index]
        let caption = big ? featText.title : album.name
        let ps = NSMutableParagraphStyle()
        ps.alignment = .center
        ps.lineBreakMode = .byTruncatingTail
        (caption as NSString).draw(
            in: NSRect(x: card.minX + 4, y: card.minY + h * 0.045,
                       width: card.width - 8, height: h * 0.11),
            withAttributes: [.font: hand(max(9, w * 0.10)),
                             .foregroundColor: NSColor(red: 0.18, green: 0.20, blue: 0.32, alpha: 1),
                             .paragraphStyle: ps])

        // Pin.
        let pinR = w * 0.045
        let pin = NSPoint(x: 0, y: card.maxY - pinR * 1.6)
        NSColor(white: 0, alpha: 0.3).setFill()
        NSBezierPath(ovalIn: NSRect(x: pin.x - pinR * 0.8, y: pin.y - pinR * 1.2,
                                    width: pinR * 1.6, height: pinR * 1.6)).fill()
        NSGradient(starting: pal.accent, ending: pal.accent.withBrightness(0.35))?
            .draw(in: NSBezierPath(ovalIn: NSRect(x: pin.x - pinR, y: pin.y - pinR,
                                                  width: pinR * 2, height: pinR * 2)),
                  angle: -60)
        NSColor(white: 1, alpha: 0.55).setFill()
        NSBezierPath(ovalIn: NSRect(x: pin.x - pinR * 0.35, y: pin.y + pinR * 0.05,
                                    width: pinR * 0.5, height: pinR * 0.5)).fill()
        ctx.restoreGState()
    }
}

File: windows/src/AlbumCoverScreenSaver.Saver/PolaroidRenderer.cs
using AlbumCoverScreenSaver.Shared;
using SkiaSharp;

namespace AlbumCoverScreenSaver.Saver;

/// <summary>
/// Polaroid Corkboard. Instant photographs pinned to a cork board in a wooden
/// frame, one in the middle for whatever is playing.
/// </summary>
/// <remarks>
/// <para>
/// This is the style the persistent board was built for. Every record you listen
/// to gets pinned up when the next one starts, and it stays exactly where it was
/// put: not re-scattered when the archive grows, not moved when the resolution
/// changes, not wiped when a new song is recorded. Come back an hour later and
/// the wall is a record of the hour.
/// </para>
/// <para>
/// <b>The one rule that has to be right.</b> The test for "has the middle
/// changed" reads the playing album's id, never the featured index. The featured
/// index is deliberately frozen for 1.35 seconds while a change animates, so
/// skipping two tracks quickly inside that window would make the middle one
/// vanish without ever reaching the wall. Reading the id catches every record
/// regardless of what the animation is doing.
/// </para>
/// </remarks>
internal sealed class PolaroidRenderer : IStyleRenderer, IDisposable
{
    /// <summary>Seconds a new photograph takes to land.</summary>
    private const float ArrivalSeconds = 0.6f;

    private readonly SaverData _data;
    private readonly AlbumPicker _picker;
    private readonly PaletteStore _palettes;
    private readonly Random _random;
    private readonly bool _isPreview;

    private readonly Featured _featured;
    private readonly ScatterLayout _board;
    private readonly CorkTexture _cork = new();
    private readonly SKPaint _fill = new() { IsAntialias = true };
    private readonly SKPaint _shadowPaint = new() { IsAntialias = true, FilterQuality = SKFilterQuality.Low };
    private readonly ShadowSprite _shadows = new();

    private string? _lastCentreId;
    private float _width;
    private float _height;

    public PolaroidRenderer(
        SaverData data, AlbumPicker picker, PaletteStore palettes, Random random, bool isPreview)
    {
        _data = data;
        _picker = picker;
        _palettes = palettes;
        _random = random;
        _isPreview = isPreview;
        _featured = new Featured(picker, data.Settings.RecencyBias);
        _board = new ScatterLayout(random);
    }

    private int Cap => _isPreview ? 4 : Math.Clamp(_data.Settings.PolaroidCount, 0, 24);

    public void BuildLayout(float width, float height)
    {
        _width = width;
        _height = height;

        _board.Claim(CollageMode.Polaroid);

        var albums = _data.Albums;
        if (albums.Count == 0) return;

        _featured.Reset(0, _picker.Pick(albums.Count, _data.Settings.RecencyBias));

        if (Cap == 0)
        {
            _board.TrimTo(0);
            return;
        }

        // Populated once and then only added to. An archive refresh must not
        // rebuild it, which is why this returns rather than reseeding.
        if (_board.Table.Count > 0) return;

        var centre = _data.LiveAlbumIndex ?? _featured.Index;
        var centreId = centre >= 0 && centre < albums.Count ? albums[centre].Id : "";

        var seeds = albums
            .Take(Cap + 1)
            .Where(album => !string.Equals(album.Id, centreId, StringComparison.Ordinal))
            .Take(Cap)
            .Select(album => album.Id)
            .ToList();

        _board.Seed(seeds, Cap, phase: 0, FindSpot);

        Log.Write($"polaroid layout: {_board.Table.Count} photos on a board of {width:0}x{height:0} points");
    }

    /// <summary>
    /// Somewhere free on the board, keeping clear of the middle photograph.
    /// </summary>
    private TableSpot? FindSpot() => _board.FindSpot(
        _width, _height,
        minSizeFraction: 0.17f, maxSizeFraction: 0.26f, scale: 1f,
        keepOutCentreX: _width * 0.5f,
        keepOutCentreY: _height * 0.5f,
        keepOutRadius: Math.Min(_height * 0.36f, _width * 0.26f) * 0.82f);

    public void Advance(double phase, float width, float height)
    {
        _width = width;
        _height = height;

        var albums = _data.Albums;
        if (albums.Count == 0) return;

        _featured.Advance(phase, _data.LiveAlbumIndex, _data.Settings.FeatureSeconds, albums.Count);

        if (_board.Table.Count == 0 && Cap > 0) BuildLayout(width, height);

        // Read off the playback signal, not the featured index. See the note at
        // the top: the index is frozen mid-change and a quickly skipped track
        // would never reach the wall.
        string? currentId = null;

        if (_data.NowPlaying.IsLive && !string.IsNullOrEmpty(_data.NowPlaying.AlbumId))
        {
            currentId = _data.NowPlaying.AlbumId;
        }
        else if (_featured.Index >= 0 && _featured.Index < albums.Count)
        {
            currentId = albums[_featured.Index].Id;
        }

        if (currentId is null) return;

        if (_lastCentreId is not null && !string.Equals(_lastCentreId, currentId, StringComparison.Ordinal))
        {
            _board.Pin(_lastCentreId, FindSpot(), phase);
            _board.TrimTo(Cap);
        }

        _lastCentreId = currentId;
    }

    public void Draw(SKCanvas canvas, float width, float height, double phase)
    {
        var albums = _data.Albums;
        if (albums.Count == 0) return;

        var centre = Math.Clamp(_featured.Index, 0, albums.Count - 1);
        var palette = _palettes.For(albums[centre].Id, _data.ImageFor(centre));

        DrawFrame(canvas, width, height);
        DrawBoard(canvas, width, height);

        // Oldest first, so newer photographs overlap the ones already up.
        for (var i = 0; i < _board.Table.Count; i++)
        {
            var pinned = _board.Table[i];
            var index = IndexOf(pinned.AlbumId);
            if (index < 0) continue;

            var settle = (float)(phase - pinned.AddedAt);
            var landed = settle < ArrivalSeconds ? Ease.Back(settle / ArrivalSeconds) : 1f;

            // It drops the last fraction of a screen into place and shrinks from
            // 120% as it goes, which is what makes it read as being put there
            // rather than appearing.
            var centreX = pinned.Nx * width;
            var centreY = (pinned.Ny * height) - ((1f - landed) * height * 0.025f);
            var side = pinned.SizeFraction * height * (1.2f - (0.2f * landed));

            DrawPhoto(
                canvas, index, centreX, centreY, side, pinned.Angle, landed,
                i, big: false, phase, palette);
        }

        var centreSide = Math.Min(height * 0.36f, width * 0.26f);
        DrawPhoto(
            canvas, centre, width * 0.5f, height * 0.5f, centreSide, -0.03f, 1f,
            index: 999, big: true, phase, palette);
    }

    private int IndexOf(string albumId) =>
        _data.AlbumIndexById.TryGetValue(albumId, out var index) ? index : -1;

    /// <summary>The wooden surround, with its grain.</summary>
    private void DrawFrame(SKCanvas canvas, float width, float height)
    {
        using var shader = SKShader.CreateLinearGradient(
            new SKPoint(0, 0),
            new SKPoint(width * 0.342f, height),
            new[] { new SKColor(107, 69, 36), new SKColor(66, 41, 20) },
            SKShaderTileMode.Clamp);

        _fill.Shader = shader;
        _fill.Color = SKColors.White;
        canvas.DrawRect(SKRect.Create(0, 0, width, height), _fill);
        _fill.Shader = null;

        for (var i = 0; i < 40; i++)
        {
            _fill.Color = SKColors.Black.WithAlpha((byte)((0.05f + (i % 3 * 0.02f)) * 255f));
            canvas.DrawRect(SKRect.Create(0, i / 40f * height, width, 1.5f), _fill);
        }
    }

    /// <summary>
    /// The cork, and the shadow the frame casts onto it.
    /// </summary>
    /// <remarks>
    /// The four shadow strips are what make the board read as recessed into the
    /// frame rather than as a rectangle printed on it. They are not decoration.
    /// </remarks>
    private void DrawBoard(SKCanvas canvas, float width, float height)
    {
        var inset = Math.Min(width, height) * 0.028f;
        var board = SKRect.Create(inset, inset, width - (inset * 2f), height - (inset * 2f));

        canvas.Save();
        canvas.ClipRect(board);

        var cork = _cork.Get(board.Width, board.Height, _random);
        if (cork is not null)
        {
            _fill.Shader = null;
            _fill.Color = SKColors.White;
            canvas.DrawBitmap(cork, board, _fill);
        }

        var depth = inset * 0.9f;
        DrawEdgeShadow(canvas, SKRect.Create(board.Left, board.Top, board.Width, depth), 0, 1);
        DrawEdgeShadow(canvas, SKRect.Create(board.Left, board.Bottom - depth, board.Width, depth), 0, -1);
        DrawEdgeShadow(canvas, SKRect.Create(board.Left, board.Top, depth, board.Height), 1, 0);
        DrawEdgeShadow(canvas, SKRect.Create(board.Right - depth, board.Top, depth, board.Height), -1, 0);

        canvas.Restore();

        _fill.Shader = null;
        _fill.Color = SKColors.Black.WithAlpha(89);
        _fill.Style = SKPaintStyle.Stroke;
        _fill.StrokeWidth = 2f;
        canvas.DrawRect(board, _fill);
        _fill.Style = SKPaintStyle.Fill;
    }

    /// <summary>One strip of the recess shadow, dark at the frame and clear inward.</summary>
    private void DrawEdgeShadow(SKCanvas canvas, SKRect strip, int inwardX, int inwardY)
    {
        var from = new SKPoint(
            inwardX >= 0 ? strip.Left : strip.Right,
            inwardY >= 0 ? strip.Top : strip.Bottom);

        var to = new SKPoint(
            inwardX > 0 ? strip.Right : inwardX < 0 ? strip.Left : from.X,
            inwardY > 0 ? strip.Bottom : inwardY < 0 ? strip.Top : from.Y);

        using var shader = SKShader.CreateLinearGradient(
            from, to,
            new[] { SKColors.Black.WithAlpha(77), SKColors.Black.WithAlpha(0) },
            SKShaderTileMode.Clamp);

        _fill.Shader = shader;
        _fill.Color = SKColors.White;
        canvas.DrawRect(strip, _fill);
        _fill.Shader = null;
    }

    /// <summary>
    /// One instant photograph: card, image, caption and pin.
    /// </summary>
    /// <remarks>
    /// It rocks rather than slides, because it turns about the pin rather than
    /// about its own centre. That is the whole difference between a photograph
    /// hanging on a wall and a sticker sliding across one.
    /// </remarks>
    private void DrawPhoto(
        SKCanvas canvas, int album, float centreX, float centreY, float side,
        float angle, float alpha, int index, bool big, double phase, ArtPalette palette)
    {
        if (alpha <= 0.01f) return;

        var w = side;
        var border = w * 0.055f;
        var h = w + (border * 4.4f);

        // Every photograph sways on its own phase, so the wall breathes instead
        // of rocking in time.
        var sway = MathF.Sin(((float)phase * 0.6f) + (index * 1.7f)) * 0.012f;

        canvas.Save();
        canvas.Translate(centreX, centreY - (h * 0.42f));
        canvas.RotateRadians(angle + sway);
        canvas.Translate(0, h * 0.42f);

        var card = SKRect.Create(-(w / 2f) - border, -h / 2f, w + (border * 2f), h);

        var sprite = _shadows.Get(0.01f);
        _shadowPaint.Color = SKColors.White.WithAlpha((byte)Math.Clamp(0.55f * alpha * 255f, 0f, 255f));
        canvas.DrawBitmap(sprite, ShadowSprite.Placement(card), _shadowPaint);

        _fill.Shader = null;
        _fill.Color = new SKColor(247, 245, 237, (byte)Math.Clamp(alpha * 255f, 0f, 255f));
        canvas.DrawRect(card, _fill);

        var photo = SKRect.Create(-w / 2f, card.Top + border, w, w);
        Drawing.DrawImage(canvas, _data.ImageFor(album), photo, alpha, radius: 0f, shadow: false);

        // Only the pinned ones are knocked back. Leaving the middle one bright
        // is what says it is the record playing.
        if (!big)
        {
            _fill.Color = palette.Deep.ToSkia(0.20f * alpha);
            canvas.DrawRect(photo, _fill);
        }

        DrawCaption(canvas, album, card, w, h, big, alpha);
        DrawPin(canvas, card, w, alpha, palette);

        canvas.Restore();
    }

    private void DrawCaption(
        SKCanvas canvas, int album, SKRect card, float w, float h, bool big, float alpha)
    {
        var albums = _data.Albums;
        if (album < 0 || album >= albums.Count) return;

        var caption = big ? FeaturedText.For(_data, album).Title : albums[album].Name;
        if (string.IsNullOrEmpty(caption)) return;

        using var paint = new SKPaint
        {
            IsAntialias = true,
            TextSize = Math.Max(9f, w * 0.10f),
            Color = new SKColor(46, 51, 82, (byte)Math.Clamp(alpha * 255f, 0f, 255f)),
            TextAlign = SKTextAlign.Center,
            // A hand on a photograph, not a typeface. The Mac uses Bradley Hand
            // and this is the nearest thing Windows ships.
            Typeface = SKTypeface.FromFamilyName(
                           "Segoe Script", SKFontStyleWeight.Normal,
                           SKFontStyleWidth.Normal, SKFontStyleSlant.Upright)
                       ?? TextLayout.Face(bold: false),
        };

        // Clipped rather than wrapped, and truncated rather than shrunk: a long
        // album name written on a photograph runs out of room, which is what a
        // person's hand does too.
        var box = SKRect.Create(card.Left + 4f, card.Top + (h * 0.845f), card.Width - 8f, h * 0.11f);

        canvas.Save();
        canvas.ClipRect(box);
        canvas.DrawText(caption, box.MidX, box.Top + paint.TextSize, paint);
        canvas.Restore();
    }

    /// <summary>
    /// The drawing pin. The only part of a photograph that takes the album's
    /// colour.
    /// </summary>
    private void DrawPin(SKCanvas canvas, SKRect card, float w, float alpha, ArtPalette palette)
    {
        var radius = w * 0.045f;
        var x = 0f;
        var y = card.Top + (radius * 1.6f);

        var opacity = (byte)Math.Clamp(alpha * 255f, 0f, 255f);

        _fill.Shader = null;
        _fill.Color = SKColors.Black.WithAlpha((byte)Math.Clamp(0.30f * alpha * 255f, 0f, 255f));
        canvas.DrawCircle(x, y + (radius * 0.4f), radius * 0.8f, _fill);

        using (var shader = SKShader.CreateLinearGradient(
                   new SKPoint(x - radius, y - radius),
                   new SKPoint(x + radius, y + radius),
                   new[]
                   {
                       palette.Accent.ToSkia().WithAlpha(opacity),
                       palette.Accent.WithBrightness(0.35f).ToSkia().WithAlpha(opacity),
                   },
                   SKShaderTileMode.Clamp))
        {
            _fill.Shader = shader;
            _fill.Color = SKColors.White;
            canvas.DrawCircle(x, y, radius, _fill);
            _fill.Shader = null;
        }

        _fill.Color = SKColors.White.WithAlpha((byte)Math.Clamp(0.55f * alpha * 255f, 0f, 255f));
        canvas.DrawCircle(x - (radius * 0.1f), y - (radius * 0.3f), radius * 0.25f, _fill);
    }

    public void Dispose()
    {
        _fill.Dispose();
        _shadowPaint.Dispose();
        _shadows.Dispose();
        _cork.Dispose();
    }
}
